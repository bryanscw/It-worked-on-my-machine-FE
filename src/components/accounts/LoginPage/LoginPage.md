### Example:

```js
import MaterialTable from 'material-table';
import { Router, withRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { withReduxRouter } from '../../../utils/mock.js';
const initialState = {
}
withReduxRouter(<LoginPage/>, initialState).renderedComponent    

```