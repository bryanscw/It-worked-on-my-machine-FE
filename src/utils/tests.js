import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import rootReducer from '../redux/rootReducer';

export const renderWithRedux = (component, initialState = {}) => {
    const store = createStore(
        rootReducer,
        initialState,
        applyMiddleware(thunk),
    )
    return <Provider store={store}>{component}</Provider>;
}

export const renderWithRouter = (component, route = '/') => {
    const history = createMemoryHistory({ initialEntries: [route] });
    return <Router history={history}>{component}</Router>
}

export const renderWithReduxRouter = (component, initialState = {}, route = '/') => {
    return renderWithRedux(renderWithRouter(component, route), initialState);
}