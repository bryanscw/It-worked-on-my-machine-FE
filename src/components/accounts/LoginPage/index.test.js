import React from 'react'
import { render, cleanup, fireEvent, screen } from '@testing-library/react'
import LoginPage from './index.js';
import { renderWithReduxRouter } from '../../../utils/tests.js';

describe('Unit Testing for Login Page', () => {

afterEach(() => {
    cleanup();
});

const delay = ms => new Promise(res => setTimeout(res, ms));

it('should take a snapshot', () => {
    const { asFragment } = renderWithReduxRouter(<LoginPage />);
    expect(asFragment()).toMatchSnapshot();
})

it('allows the user to login successfully', async () => {
  const fakeUserResponse = {access_token: 'fake_user_token'}
  jest.spyOn(window, 'fetch').mockImplementationOnce(() => {
    return Promise.resolve({
      json: () => Promise.resolve(fakeUserResponse),
    })
  })
  renderWithReduxRouter(<LoginPage />)
  fireEvent.change(screen.getByTestId("usernameField").querySelector('input'), {
    target: {value: 'admin1@test.com'},
  })
  fireEvent.change(screen.getByTestId("passwordField").querySelector('input'), {
    target: {value: 'admin123'},
  })
  fireEvent.click(screen.getByTestId("loginButton"))
  await delay(1000)
  expect(window.localStorage.getItem('access_token')).toEqual(fakeUserResponse.access_token)
  window.localStorage.removeItem('access_token')
})

it('wrong password reject user login ', async () => {
  jest.spyOn(window, 'fetch').mockImplementationOnce(() => {
    return Promise.resolve({
      json: () => Promise.reject(),
    })
  })
  renderWithReduxRouter(<LoginPage />)
  fireEvent.change(screen.getByTestId("usernameField").querySelector('input'), {
    target: {value: 'admin1@test.com'},
  })
  fireEvent.change(screen.getByTestId("passwordField").querySelector('input'), {
    target: {value: 'wrong password'},
  })
  fireEvent.click(screen.getByTestId("loginButton"))
  await delay(1000)
  expect(window.localStorage.getItem('access_token')).toEqual(null)
})

})