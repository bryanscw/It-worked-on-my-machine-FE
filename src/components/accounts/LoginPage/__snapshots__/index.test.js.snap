// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit Testing for Login Page should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="LoginPage-main-1"
    data-test="loginContainer"
  >
    <main
      class="MuiContainer-root MuiContainer-maxWidthXs"
    >
      <img
        alt="Mazematic Icon"
        class="LoginPage-icon-8"
        src="mazematicsIcon.png"
      />
      <div
        class="LoginPage-paper-4"
      >
        <img
          alt="Pokemon Icon"
          class="LoginPage-avatar-5"
          src="pokemonIcon.png"
        />
        <form
          class="LoginPage-form-6"
          novalidate=""
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
            data-testid="usernameField"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-focused Mui-focused Mui-required Mui-required"
              data-shrink="true"
              for="username"
              id="username-label"
            >
              Username
              <span
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth Mui-focused Mui-focused MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                autocomplete="username"
                class="MuiInputBase-input MuiOutlinedInput-input"
                id="username"
                name="username"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-75 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-77 PrivateNotchedOutline-legendNotched-78"
                >
                  <span>
                    Username *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
            data-testid="passwordField"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
              data-shrink="false"
              for="password"
              id="password-label"
            >
              Password
              <span
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                autocomplete="current-password"
                class="MuiInputBase-input MuiOutlinedInput-input"
                id="password"
                name="password"
                required=""
                type="password"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-75 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-77"
                >
                  <span>
                    Password *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained LoginPage-submit-7 MuiButton-containedPrimary MuiButton-fullWidth"
            data-testid="loginButton"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Login
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </form>
      </div>
    </main>
  </div>
</DocumentFragment>
`;

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="LoginPage-main-1"
    data-test="loginContainer"
  >
    <main
      class="MuiContainer-root MuiContainer-maxWidthXs"
    >
      <img
        alt="Mazematic Icon"
        class="LoginPage-icon-8"
        src="mazematicsIcon.png"
      />
      <div
        class="LoginPage-paper-4"
      >
        <img
          alt="Pokemon Icon"
          class="LoginPage-avatar-5"
          src="pokemonIcon.png"
        />
        <form
          class="LoginPage-form-6"
          novalidate=""
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
            data-testid="usernameField"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-focused Mui-focused Mui-required Mui-required"
              data-shrink="true"
              for="username"
              id="username-label"
            >
              Username
              <span
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth Mui-focused Mui-focused MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                autocomplete="username"
                class="MuiInputBase-input MuiOutlinedInput-input"
                id="username"
                name="username"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-75 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-77 PrivateNotchedOutline-legendNotched-78"
                >
                  <span>
                    Username *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
            data-testid="passwordField"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
              data-shrink="false"
              for="password"
              id="password-label"
            >
              Password
              <span
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                autocomplete="current-password"
                class="MuiInputBase-input MuiOutlinedInput-input"
                id="password"
                name="password"
                required=""
                type="password"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-75 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-77"
                >
                  <span>
                    Password *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained LoginPage-submit-7 MuiButton-containedPrimary MuiButton-fullWidth"
            data-testid="loginButton"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Login
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </form>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
